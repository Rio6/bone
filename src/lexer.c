#include "lexer.h"
#include "stream.h"

#include <stdlib.h>
#include <string.h>
#include <ctype.h>

static Lexer * const lexers[] = {
   comment_lexer,
   number_lexer,
   string_lexer,
   ident_lexer,
   sep_lexer,
   assign_lexer,
   operator_lexer,
};

Token *lex(Stream *stream) {
   Token head_token = {0};
   Token *last_token = &head_token;

   while(1) {
      int has_token = 0;
      char last_c = '\0';

      for(size_t i = 0; i < sizeof(lexers) / sizeof(lexers[0]); i++) {
         // advance until not a white space
         while(isspace(stream_getc(stream)));
         last_c = stream_ungetc(stream, 1);

         Token *token = lexers[i](stream);
         if(token) {
            last_token->next = token;
            last_token = token;
            has_token = 1;
            break;
         }
      }

      if(!has_token) {
         if(last_c != EOF) {
            fprintf(stderr, "Unexpected character: %1$c (%1$X)\n", last_c);
         }
         break;
      }
   }

   return head_token.next;
}

Token *token_create(TokenType type, const char *value) {
   Token *token = calloc(1, sizeof(Token));

   token->type = type;

   if(value) {
      size_t len = strlen(value) + 1;
      token->value = malloc(len);
      memcpy(token->value, value, len);
   }

   return token;
}

void token_delete(Token *token) {
   if(!token) return;
   token_delete(token->next);
   free(token->value);
   free(token);
}

void token_dump(Token *token) {
   if(!token) return;
   printf("type: %s, value: %s\n", token_type_name(token->type), token->value);
   token_dump(token->next);
}

char *token_type_name(TokenType type) {
   switch(type) {
      case IDENT:   return "IDENT";
      case KEYWORD: return "KEYWORD";
      case INT:     return "INT";
      case FLOAT:   return "FLOAT";
      case CHAR:    return "CHAR";
      case STRING:  return "STRING";
      case COMMENT: return "COMMENT";
      case BLOCK_L: return "BLOCK_L";
      case BLOCK_R: return "BLOCK_R";
      case PAREN_L: return "PAREN_L";
      case PAREN_R: return "PAREN_R";
      case LIST_L:  return "LIST_L";
      case LIST_R:  return "LIST_R";
      case DOT:     return "DOT";
      case COMMA:   return "COMMA";
      case SEMICOLON: return "SEMICOLON";
      case ASSIGN:  return "ASSIGN";
      case OP:      return "OP";
      case REF:     return "REF";
      case DEREF:   return "DEREF";
      default:      return "UNKNOWN";
   }
}
