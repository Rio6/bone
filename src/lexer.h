#pragma once

#include "stream.h"

typedef enum {
   ERROR,
   IDENT,
   KEYWORD,
   INT,
   FLOAT,
   CHAR,
   STRING,
   COMMENT,
   BLOCK_L,
   BLOCK_R,
   PAREN_L,
   PAREN_R,
   LIST_L,
   LIST_R,
   DOT,
   COMMA,
   SEMICOLON,
   ASSIGN,
   OP,
   REF,
   DEREF,
} TokenType;

typedef struct Token {
   TokenType type;
   char *value; // owner
   struct Token *next; // owner
} Token;

typedef Token *(Lexer)(Stream *stream);

extern Lexer *lexers[];
extern const char *TokenType_names[];

Token *lex(Stream *stream);

Token *token_create(TokenType type, const char *value);
Token *token_create_raw(TokenType type, char *value);
void token_delete(Token*);
void token_dump(Token*);

char *token_type_name(TokenType);

Token *ident_lexer(Stream*);
Token *sep_lexer(Stream*);
Token *number_lexer(Stream*);
Token *char_lexer(Stream*);
Token *string_lexer(Stream*);
Token *comment_lexer(Stream*);
Token *assign_lexer(Stream*);
Token *operator_lexer(Stream*);
