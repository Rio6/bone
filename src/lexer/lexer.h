#pragma once

#include "token.h"
#include "utils/stream.h"

typedef Token *(Lexer)(Stream *stream);

extern Lexer *lexers[];

Token *lexer_run(Stream *stream);
