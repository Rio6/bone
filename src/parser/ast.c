#include "ast.h"
#include "utils/oop.h"

#include <stdio.h>

const char *ASTType_names[] = {
   "AST_ERROR",
   "AST_ATOM",
   "AST_PREFIX",
   "AST_POSTFIX",
   "AST_BINARY",
   "AST_TENARY",
   "AST_GROUP",
   "AST_TOKEN",
};

void print_indent(unsigned indent) {
   for(unsigned i = 0; i < indent; i++) {
      fputs("   ", stdout);
   }
}

void ast_init(ASTNode *node, ASTType type, ASTParseFn parse, ASTDumpFn dump, ASTDeleteFn delete) {
   node->type = type;
   node->parse = parse;
   node->dump = dump;
   node->delete = delete;
}

ASTNode *ast_root(ASTNode *node) {
   while(node->up) node = node->up;
   return node;
}
