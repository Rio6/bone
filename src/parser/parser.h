#pragma once

#include "lexer/token.h"

// Forward typedef
typedef struct ASTToken ASTToken;
typedef struct ASTNode ASTNode;

typedef ASTNode *(ASTParser)(ASTToken*);

extern ASTParser **parsers[];

ASTNode *parser_run(Token*);

ASTNode *comment_parser(ASTToken*);
