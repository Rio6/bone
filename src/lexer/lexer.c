#include "lexer.h"
#include "token.h"

#include "utils/stream.h"

#include <string.h>
#include <ctype.h>

Token *lexer_run(Stream *stream) {
   Token head_token = {0};
   Token *last_token = &head_token;

   while(1) {
      int has_token = 0;
      char last_c = '\0';

      for(Lexer **lexer = lexers; *lexer != NULL; lexer++) {

         // advance until not a white space
         while(isspace(stream_getc(stream)));
         last_c = stream_ungetc(stream, 1);

         Token *token = (*lexer)(stream);

         if(token) {
            last_token->next = token;
            last_token = token;
            has_token = 1;
            break;
         }
      }

      if(!has_token) {
         char buff[128];
         snprintf(buff, sizeof(buff), "Unexpected character: %1$c (%1$X)\n", last_c);
         last_token->next = token_create(T_ERROR, buff);
         last_token = last_token->next;
      }

      if(last_token->type == T_ERROR || last_token->type == T_EOT) {
         break;
      }
   }

   return head_token.next;
}
