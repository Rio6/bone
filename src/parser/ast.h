#pragma once

#include "parser.h"

typedef enum {
   ATOM,
   UNARY,
   BINARY,
   TENARY,
   GROUP,
   TOKEN,
} ASTType;
extern const char *ASTType_names[];

typedef struct ASTNode ASTNode;
struct ASTNode {
   ASTType type;
   ASTNode *up;
   ASTNode *prev;
   ASTNode *next;

   ASTNode *(*parse)(ASTNode*, ASTParser*);
   void (*dump)(ASTNode*, unsigned indent);
   void (*delete)(ASTNode*);
};

void print_indent(unsigned);
ASTNode *ast_root(ASTNode*);
