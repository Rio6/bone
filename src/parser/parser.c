#include "parser.h"
#include "ast.h"

#include "ast_token.h"
#include "ast_error.h"
#include "ast_group.h"

#include "utils/oop.h"

#include <stddef.h>

ASTParser **parsers[] = {
   (ASTParser*[]) { group_parser, NULL },
   NULL,
};

ASTNode *parser_run(Token *tokens) {
   ASTNode *ast = (ASTNode*) ast_token_from_tokens(tokens);

   for(ASTParser ***parser = parsers; *parser != NULL; parser++) {
      ast = ast_root(CALL_METHOD(parse, ast, *parser));
   }

   return ast;
}
