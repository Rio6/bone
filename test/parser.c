#include "lexer/lexer.h"

#include "parser/ast.h"
#include "parser/parser.h"

#include "utils/stream.h"
#include "utils/oop.h"

#include <stdio.h>

#define TEST_FILE "../idea.bone"

int main() {
   FILE *file = fopen(TEST_FILE, "r");

   if(!file) {
      perror("error open file");
      return 1;
   }

   Stream stream = {
      .stream = file
   };

   Token *tokens = lexer_run(&stream);
   fclose(file);

   ASTNode *ast = parser_run(tokens);

   CALL_METHOD(dump, ast, 0);
   CALL_METHOD(delete, ast);

   return 0;
}

