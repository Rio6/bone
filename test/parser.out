TOKEN: type: KEYWORD, value: import
TOKEN: type: IDENT, value: std
TOKEN: type: DOT, value: (null)
TOKEN: type: IDENT, value: math
TOKEN: type: DOT, value: (null)
TOKEN: type: SEMICOLON, value: (null)
TOKEN: type: KEYWORD, value: import
TOKEN: type: IDENT, value: std
TOKEN: type: DOT, value: (null)
TOKEN: type: IDENT, value: io
TOKEN: type: SEMICOLON, value: (null)
TOKEN: type: KEYWORD, value: import
TOKEN: type: IDENT, value: directory
TOKEN: type: DOT, value: (null)
TOKEN: type: IDENT, value: file
TOKEN: type: SEMICOLON, value: (null)
TOKEN: type: KEYWORD, value: import
TOKEN: type: IDENT, value: directory
TOKEN: type: DOT, value: (null)
TOKEN: type: STRING, value: fancy-name
TOKEN: type: DOT, value: (null)
TOKEN: type: IDENT, value: name
TOKEN: type: SEMICOLON, value: (null)
TOKEN: type: KEYWORD, value: export
TOKEN: type: KEYWORD, value: struct
TOKEN: type: IDENT, value: A
GROUP: BLOCK
  TOKEN: type: IDENT, value: i32
  TOKEN: type: IDENT, value: a
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: COMMENT, value: (null)
  TOKEN: type: IDENT, value: i32
  TOKEN: type: IDENT, value: b
  TOKEN: type: OP, value: =
  TOKEN: type: INT, value: 10
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: f64
  TOKEN: type: IDENT, value: c
  TOKEN: type: OP, value: =
  TOKEN: type: FLOAT, value: 10.23
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: f32
  TOKEN: type: IDENT, value: d
  TOKEN: type: OP, value: =
  TOKEN: type: FLOAT, value: .65
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: f32
  TOKEN: type: IDENT, value: e
  TOKEN: type: OP, value: =
  TOKEN: type: FLOAT, value: .4e10
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: f32
  TOKEN: type: IDENT, value: f
  TOKEN: type: OP, value: =
  TOKEN: type: FLOAT, value: 4e4
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: char
  GROUP: PAREN
    TOKEN: type: IDENT, value: A
    TOKEN: type: IDENT, value: self
    TOKEN: type: COMMA, value: (null)
    TOKEN: type: IDENT, value: bool
    TOKEN: type: IDENT, value: stuff
  TOKEN: type: REF, value: (null)
  TOKEN: type: IDENT, value: method
  TOKEN: type: OP, value: =
  GROUP: BLOCK
    TOKEN: type: KEYWORD, value: return
    TOKEN: type: CHAR, value: x
    TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: char
  GROUP: PAREN
    TOKEN: type: IDENT, value: A
    TOKEN: type: COMMA, value: (null)
    TOKEN: type: IDENT, value: bool
  TOKEN: type: REF, value: (null)
  TOKEN: type: IDENT, value: sameThing
  TOKEN: type: OP, value: =
  TOKEN: type: IDENT, value: char
  GROUP: PAREN
    TOKEN: type: IDENT, value: A
    TOKEN: type: IDENT, value: self
    TOKEN: type: COMMA, value: (null)
    TOKEN: type: IDENT, value: bool
    TOKEN: type: IDENT, value: stuff
  GROUP: BLOCK
  TOKEN: type: SEMICOLON, value: (null)
TOKEN: type: IDENT, value: i32
TOKEN: type: REF, value: (null)
TOKEN: type: IDENT, value: aPtr
TOKEN: type: SEMICOLON, value: (null)
TOKEN: type: IDENT, value: i32
GROUP: LIST
TOKEN: type: IDENT, value: array
TOKEN: type: OP, value: =
GROUP: BLOCK
  TOKEN: type: INT, value: 1
  TOKEN: type: COMMA, value: (null)
  TOKEN: type: INT, value: 2
  TOKEN: type: COMMA, value: (null)
  TOKEN: type: INT, value: 3
  TOKEN: type: COMMA, value: (null)
  TOKEN: type: INT, value: 4
TOKEN: type: SEMICOLON, value: (null)
TOKEN: type: IDENT, value: i32
GROUP: LIST
TOKEN: type: IDENT, value: arrayAgain
TOKEN: type: OP, value: =
TOKEN: type: IDENT, value: i32
GROUP: LIST
GROUP: BLOCK
  TOKEN: type: INT, value: 1
  TOKEN: type: COMMA, value: (null)
  TOKEN: type: INT, value: 2
  TOKEN: type: COMMA, value: (null)
  TOKEN: type: INT, value: 3
  TOKEN: type: COMMA, value: (null)
  TOKEN: type: INT, value: 4
TOKEN: type: SEMICOLON, value: (null)
TOKEN: type: IDENT, value: i32
GROUP: LIST
TOKEN: type: REF, value: (null)
TOKEN: type: IDENT, value: ptrToArray
TOKEN: type: OP, value: =
TOKEN: type: REF, value: (null)
TOKEN: type: IDENT, value: array
TOKEN: type: SEMICOLON, value: (null)
TOKEN: type: IDENT, value: i32
TOKEN: type: REF, value: (null)
GROUP: LIST
TOKEN: type: IDENT, value: arrayOfPtr
TOKEN: type: OP, value: =
TOKEN: type: IDENT, value: null
TOKEN: type: SEMICOLON, value: (null)
TOKEN: type: IDENT, value: i32
GROUP: PAREN
  TOKEN: type: IDENT, value: char
TOKEN: type: REF, value: (null)
TOKEN: type: IDENT, value: ptrToFuncReturnsI32
TOKEN: type: SEMICOLON, value: (null)
TOKEN: type: COMMENT, value: (null)
TOKEN: type: IDENT, value: char
GROUP: PAREN
  TOKEN: type: IDENT, value: A
  TOKEN: type: IDENT, value: byValue
  TOKEN: type: COMMA, value: (null)
  TOKEN: type: IDENT, value: A
  TOKEN: type: REF, value: (null)
  TOKEN: type: IDENT, value: byPointer
  TOKEN: type: COMMA, value: (null)
  TOKEN: type: IDENT, value: A
  GROUP: LIST
  TOKEN: type: IDENT, value: byValue
  TOKEN: type: COMMA, value: (null)
  TOKEN: type: IDENT, value: A
  GROUP: LIST
  TOKEN: type: REF, value: (null)
  TOKEN: type: IDENT, value: byReferenceOfArray
TOKEN: type: IDENT, value: aFunction
TOKEN: type: OP, value: =
GROUP: BLOCK
  TOKEN: type: KEYWORD, value: if
  GROUP: PAREN
    TOKEN: type: IDENT, value: byValue
    TOKEN: type: DOT, value: (null)
    TOKEN: type: IDENT, value: a
    TOKEN: type: OP, value: ==
    TOKEN: type: IDENT, value: byPointer
    TOKEN: type: DOT, value: (null)
    TOKEN: type: IDENT, value: b
  GROUP: BLOCK
    TOKEN: type: COMMENT, value: (null)
    TOKEN: type: KEYWORD, value: return
    TOKEN: type: IDENT, value: byPointer
    TOKEN: type: DOT, value: (null)
    TOKEN: type: IDENT, value: c
    TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: A
  TOKEN: type: IDENT, value: copied
  TOKEN: type: OP, value: =
  TOKEN: type: DEREF, value: (null)
  TOKEN: type: IDENT, value: byPointer
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: A
  TOKEN: type: IDENT, value: object
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: COMMENT, value: (null)
  TOKEN: type: IDENT, value: A
  TOKEN: type: IDENT, value: another
  TOKEN: type: OP, value: =
  GROUP: BLOCK
    TOKEN: type: INT, value: 1
    TOKEN: type: COMMA, value: (null)
    TOKEN: type: COMMENT, value: (null)
    TOKEN: type: IDENT, value: c
    TOKEN: type: OP, value: =
    TOKEN: type: INT, value: 10
    TOKEN: type: COMMA, value: (null)
    TOKEN: type: INT, value: 123
    TOKEN: type: COMMA, value: (null)
    TOKEN: type: COMMENT, value: (null)
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: KEYWORD, value: return
  TOKEN: type: CHAR, value: x
  TOKEN: type: SEMICOLON, value: (null)
TOKEN: type: IDENT, value: void
GROUP: PAREN
TOKEN: type: IDENT, value: normalFunction
TOKEN: type: OP, value: =
GROUP: BLOCK
  TOKEN: type: IDENT, value: io
  TOKEN: type: DOT, value: (null)
  TOKEN: type: IDENT, value: print
  GROUP: PAREN
    TOKEN: type: IDENT, value: stuff
  TOKEN: type: SEMICOLON, value: (null)
TOKEN: type: IDENT, value: i32
GROUP: PAREN
TOKEN: type: OP, value: =
GROUP: BLOCK
  TOKEN: type: IDENT, value: i32
  TOKEN: type: IDENT, value: a
  TOKEN: type: OP, value: =
  TOKEN: type: INT, value: 10
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: KEYWORD, value: while
  GROUP: PAREN
    TOKEN: type: IDENT, value: a
    TOKEN: type: OP, value: <
    TOKEN: type: INT, value: 0
  GROUP: BLOCK
    TOKEN: type: IDENT, value: a
    TOKEN: type: OP, value: <<=
    TOKEN: type: INT, value: 1
    TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: KEYWORD, value: for
  GROUP: PAREN
    TOKEN: type: IDENT, value: i32
    TOKEN: type: IDENT, value: i
    TOKEN: type: OP, value: =
    TOKEN: type: INT, value: 0
    TOKEN: type: SEMICOLON, value: (null)
    TOKEN: type: IDENT, value: i
    TOKEN: type: OP, value: <
    TOKEN: type: INT, value: 100
    TOKEN: type: SEMICOLON, value: (null)
    TOKEN: type: IDENT, value: i
    TOKEN: type: OP, value: +=
    TOKEN: type: INT, value: 1
  GROUP: BLOCK
    TOKEN: type: KEYWORD, value: if
    GROUP: PAREN
      TOKEN: type: IDENT, value: i
      TOKEN: type: OP, value: ==
      TOKEN: type: INT, value: 5
    GROUP: BLOCK
      TOKEN: type: IDENT, value: continue
      TOKEN: type: SEMICOLON, value: (null)
    TOKEN: type: KEYWORD, value: if
    GROUP: PAREN
      TOKEN: type: IDENT, value: i
      TOKEN: type: OP, value: ==
      TOKEN: type: INT, value: 8
    GROUP: BLOCK
      TOKEN: type: IDENT, value: break
      TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: f32
  TOKEN: type: IDENT, value: f
  TOKEN: type: OP, value: =
  TOKEN: type: IDENT, value: sqrt
  GROUP: PAREN
    TOKEN: type: INT, value: 2
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: f32
  TOKEN: type: IDENT, value: g
  TOKEN: type: OP, value: =
  TOKEN: type: IDENT, value: sqrt
  GROUP: PAREN
    TOKEN: type: FLOAT, value: 2.5e
    TOKEN: type: OP, value: -
    TOKEN: type: INT, value: 4
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: KEYWORD, value: if
  GROUP: PAREN
    TOKEN: type: IDENT, value: true
  TOKEN: type: KEYWORD, value: if
  GROUP: PAREN
    TOKEN: type: IDENT, value: a
    TOKEN: type: OP, value: ==
    TOKEN: type: INT, value: 10
  TOKEN: type: IDENT, value: a
  TOKEN: type: OP, value: =
  TOKEN: type: INT, value: 11
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: KEYWORD, value: else
  TOKEN: type: IDENT, value: a
  TOKEN: type: OP, value: =
  TOKEN: type: INT, value: 10
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: KEYWORD, value: switch
  GROUP: PAREN
    TOKEN: type: IDENT, value: g
    TOKEN: type: OP, value: *
    TOKEN: type: FLOAT, value: 0.5
  GROUP: BLOCK
    TOKEN: type: KEYWORD, value: case
    GROUP: PAREN
      TOKEN: type: INT, value: 0
    TOKEN: type: IDENT, value: print
    GROUP: PAREN
      TOKEN: type: IDENT, value: g
    TOKEN: type: SEMICOLON, value: (null)
    TOKEN: type: KEYWORD, value: case
    GROUP: PAREN
      TOKEN: type: INT, value: 1
      TOKEN: type: COMMA, value: (null)
      TOKEN: type: INT, value: 2
      TOKEN: type: COMMA, value: (null)
      TOKEN: type: INT, value: 3
    GROUP: BLOCK
      TOKEN: type: IDENT, value: char
      TOKEN: type: IDENT, value: x
      TOKEN: type: OP, value: =
      TOKEN: type: CHAR, value: 1
      TOKEN: type: SEMICOLON, value: (null)
      TOKEN: type: IDENT, value: print
      GROUP: PAREN
        TOKEN: type: IDENT, value: x
      TOKEN: type: SEMICOLON, value: (null)
    TOKEN: type: KEYWORD, value: case
    GROUP: PAREN
    GROUP: BLOCK
      TOKEN: type: IDENT, value: print
      GROUP: PAREN
        TOKEN: type: STRING, value: default
      TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: char
  GROUP: LIST
  TOKEN: type: IDENT, value: string
  TOKEN: type: OP, value: =
  TOKEN: type: STRING, value: string $#%$^%$^$ \" escape^
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: char
  GROUP: LIST
  TOKEN: type: IDENT, value: string1
  TOKEN: type: OP, value: =
  TOKEN: type: STRING, value: concat stringishere
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: char
  GROUP: LIST
  TOKEN: type: IDENT, value: longstring
  TOKEN: type: OP, value: =
  TOKEN: type: STRING, value: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: A
  TOKEN: type: IDENT, value: a
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: IDENT, value: a
  TOKEN: type: COLON, value: (null)
  TOKEN: type: IDENT, value: method
  GROUP: PAREN
  TOKEN: type: SEMICOLON, value: (null)
  TOKEN: type: COMMENT, value: (null)
TOKEN: type: COMMENT, value: (null)
TOKEN: type: EOT, value: (null)
